syftlib/src/main/java/org/openmined/syft/SignallingClient.kt

```diff
@@ -0,0 +1,141 @@
+package org.openmined.syft
+
+import android.util.Log
+import com.google.gson.JsonObject
+import io.reactivex.BackpressureStrategy
+import io.reactivex.Flowable
+import io.reactivex.android.schedulers.AndroidSchedulers
+import io.reactivex.disposables.Disposable
+import io.reactivex.schedulers.Schedulers
+import java.io.BufferedReader
+import java.io.BufferedWriter
+import java.net.Socket
+import java.nio.charset.Charset
+import java.util.concurrent.Executors
+import java.util.concurrent.Future
+import java.util.concurrent.ScheduledFuture
+import java.util.concurrent.TimeUnit
+
+
+const val SOCKET_PING = "socket-ping"
+private const val TAG = "WebRTC Signalling Client"
+
+class SignallingClient(
+    private val signallingInterface: SignallingInterface,
+    private val workerId: String,
+    private val keepAliveTimeout: Int = 20000,
+    private val url: String,
+    private val port: Int
+) {
+    private lateinit var socket: Socket
+    private lateinit var writer: BufferedWriter
+    private lateinit var reader: BufferedReader
+
+    private val disposable: Disposable
+    private val responseObservable: Flowable<String>
+    private val keepSendingScheduler = Executors.newScheduledThreadPool(1)
+    private var sendingThread: Future<*>
+
+    init {
+        responseObservable = Flowable.create<String>(
+            { emitter ->
+                emitter.setCancellable { this.stop() }
+                try {
+                    createSocket()
+                    while (socket.isConnected) {
+                        emitter.onNext(reader.readLine())
+                    }
+                    emitter.onComplete()
+                } catch (e: Exception) {
+                    Log.e(TAG, "Socket TimeOut", e)
+                    stop()
+                    emitter.onError(e)
+                }
+            }, BackpressureStrategy.BUFFER
+        )
+
+        disposable = invokeSocketObserver()
+        sendingThread = runKeepAliveSchedule()
+    }
+
+    /**
+     * Send the data over the Socket connection to PyGrid
+     */
+    fun send(type: String, data: JsonObject) {
```

Comment:
I was thinking in using `kotlinx.serialization` for the json bit. It also includes support for protobuf

Reply:
Okay, not a big deal to shift